### PR TITLE
[cli]] updates for multi-os browser for langstream UI

### PR DESCRIPTION
**Pull request by mark878**
main issue was to figure out that the thread needs to wait a bit after processBuilder is called to start the browser and waitfor() can't be used because it doesn't return until the started process completes.

I only tested this out is WSL.  Needs to be tested on standard linux, macos and windows.

Added logic to determine OS and spawn a browser process based on the OS

### CODE REVIEW

**Comment by mark878** [comment]
updated the test(s)

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java [hunk 1]
```diff
@@ -58,17 +61,20 @@ public class UIAppCmd extends BaseApplicationCmd {
     private String applicationId;
 
     @CommandLine.Option(
-            names = {"-p", "--port"},
-            description = "Port for the local webserver and UI. If 0, a random port will be used. ",
-            defaultValue = "8092")
+        names = {
```
Comment (eolivelli) [inline langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:64]: this file has been reformatted

please run this command to make it complaint with the code style

mvn generate-sources

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java [hunk 2]
```diff
@@ -30,7 +30,10 @@
 import io.undertow.server.handlers.resource.ResourceHandler;
 import io.undertow.util.Headers;
 import io.undertow.util.HttpString;
+import java.awt.Desktop;
```
Comment (eolivelli) [inline langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:33]: this is a great idea !

I think that Desktop.browse is meant to be compatible with all the OSs
https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/awt/Desktop.html#browse(java.net.URI)

I have tested it on Mac and it works like a charm

Can you please try to use it also on your OS ?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java [hunk 3]
```diff
@@ -24,8 +25,12 @@ class UIAppCmdTest {
     @Test
     void openBrowser() {
         // ok
-        UIAppCmd.openBrowserAtPort("echo", 9029);
+	String command = "/usr/bin/echo";
+	boolean Result = UIAppCmd.checkAndLaunch( command, 80);
```
Comment (eolivelli) [inline langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java:29]: I think that we can remove checkAndLaunch given your findings on Desktop.browse()
at that point we can drop this test.

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java [hunk 4]
```diff
@@ -119,91 +125,128 @@ public static Undertow startServer(
...
+            } catch (IOException ioException) {
+                return false;
+            }
+        } else {
+            return false;
         }
     }
 
+    static boolean openBrowserAtPort(String URL, int port) {
+        String os = System.getProperty("os.name").toLowerCase();
```
Comment (nicoloboschi) [inline langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:217]: It's better to use commons-lang3 
> SystemUtils.IS_OS_MAC

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
Good work !

File: langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java (hunk 3 above)
Comment (mark878) [reply langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java:29]: Desktop.browse() didn't work in WSL linux.  Or rather it didn't when I tried it before.  That's why I went to xdg-open. 
 I can use Desktop.browse() to open the browser on a mac.  Will rewrite for mac and also try it again on WSL

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java (hunk 1 above)
Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:64]: I spent some time trying to figure what style to use.  Got educated on java style. thx will do.  BTW, I do everything at the command-line.  I use to use jetbrains -probably setup it up at some point.

Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:64]: done

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java (hunk 2 above)
Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:33]: It doesn't work in WSL linux.  I'll modify the code to launch if in a mac. For WSL linux, I tried it earlier. The WSL error was something like "can't launch headless".  Easy to try though.  Will try again.

Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:33]: Tried this.  At least I got a status message in the logs:
```
received response: (GET http://localhost:8090/api/applications/default/test?stats=false) 200
Starting UI at http://localhost:8092
Operating system identified as: linux
Could not Start browser.  Either add the proper command to your OS (open for mac or xdg-open for linux) or start a browser manually at http://localhost:8092
```

code change:
```

    static boolean openBrowserAtPort(String URL, int port) {
        String os = System.getProperty("os.name").toLowerCase();
        String openCommand = "";
        if (os.indexOf("win") >= 0
                || os.indexOf("nux") >= 0
                || os.indexOf("nix") >= 0
                || os.indexOf("nux") >= 0) {
            Desktop desktop = Desktop.getDesktop();
            try {
                desktop.browse(new URI(URL + port));
            } catch (Exception e) {
                return false;
            }
            return true;
        } else if (os.indexOf("nix") >= 0 || os.indexOf("nux") >= 0) {
            openCommand = "/usr/bin/xdg-open";
            return checkAndLaunch(openCommand, port);
        }
        return false;
```
    }

Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:33]: I will, however, have desktop browser to open the browser on a mac

File: langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java (hunk 3 above)
Comment (mark878) [reply langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java]: left checkAndLaunch in as WSL as Desktop.browse() fails on WSL linux (ubuntu 22.04).  Maybe it'll get fixed eventually.

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java [hunk 5]
```diff
@@ -15,17 +15,41 @@
...
+
+        Files.setPosixFilePermissions(
+                path, Set.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_EXECUTE));
+    }
+
     @Test
-    void openBrowser() {
+    void openBrowser() throws IOException {
+        // write test file
+        String fname = "/tmp/testfile";
```
Comment (eolivelli) [inline langstream-cli/src/test/java/ai/langstream/cli/commands/applications/UIAppCmdTest.java:42]: This test cannot pass on a Window machine.

I think that we can remove it.
Also, it is creating a fixed path, so you cannot run the same test in parallel (not a big deal, but it worth mentioning it)

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java [hunk 6]
```diff
@@ -176,19 +179,55 @@ public static Undertow startServer(
...
+        String openCommand = "";
+        if (SystemUtils.IS_OS_MAC || SystemUtils.IS_OS_WINDOWS) {
+            Desktop desktop = Desktop.getDesktop();
+            try {
+                desktop.browse(new URI(URL + port));
+            } catch (Exception e) {
+                return false;
+            }
+            return true;
+        } else if (SystemUtils.IS_OS_LINUX) {
```
Comment (eolivelli) [inline langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:226]: doesn't Desktop.browse() work on Linux ?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java (hunk 6 above)
Comment (mark878) [reply langstream-cli/src/main/java/ai/langstream/cli/commands/applications/UIAppCmd.java:226]: not in WSL.  Browser window doesn't open becasuse desktop.browse throws an execption.  

However, I can manually open a browser to the UI URL

**Review by eolivelli — APPROVED** [review]
LGTM

**Pull request merged**